src/library/scala/math/Ordering.scala

```diff
@@ -214,26 +204,59 @@ trait LowPriorityOrderingImplicits {
   * new orderings.
   */
 object Ordering extends LowPriorityOrderingImplicits {
+  private final val reverseSeed  = 41
+  private final val optionSeed   = 43
+  private final val iterableSeed = 47
+
   @inline def apply[T](implicit ord: Ordering[T]) = ord
 
+  /** A reverse ordering */
+  private final class Reverse[T](private val outer: Ordering[T]) extends Ordering[T] {
+    override def reverse: Ordering[T]       = outer
+    def compare(x: T, y: T): Int            = outer.compare(y, x)
+    override def lteq(x: T, y: T): Boolean  = outer.lteq(y, x)
+    override def gteq(x: T, y: T): Boolean  = outer.gteq(y, x)
+    override def lt(x: T, y: T): Boolean    = outer.lt(y, x)
+    override def gt(x: T, y: T): Boolean    = outer.gt(y, x)
+    override def equiv(x: T, y: T): Boolean = outer.equiv(y, x)
+    override def max[U <: T](x: U, y: U): U = outer.min(x, y)
+    override def min[U <: T](x: U, y: U): U = outer.max(x, y)
+
+    override def equals(obj: scala.Any): Boolean = obj match {
+      case that: AnyRef if this eq that => true
+      case that: Reverse[T]             => this.outer == that.outer
+      case _                            => false
+    }
+    override def hashCode(): Int = outer.hashCode() * reverseSeed
+  }
+
+  private final class IterableOrdering[CC[X] <: Iterable[X], T](private val ord: Ordering[T]) extends Ordering[CC[T]] {
+    def compare(x: CC[T], y: CC[T]): Int = {
+      val xe = x.iterator
+      val ye = y.iterator
+
+      while (xe.hasNext && ye.hasNext) {
+        val res = ord.compare(xe.next(), ye.next())
+        if (res != 0) return res
+      }
+
+      Boolean.compare(xe.hasNext, ye.hasNext)
+    }
+
+    override def equals(obj: scala.Any): Boolean = obj match {
+      case that: AnyRef if this eq that  => true
+      case that: IterableOrdering[CC, T] => this.ord == that.ord
+      case _                             => false
+    }
+    override def hashCode(): Int = ord.hashCode() * iterableSeed
+  }
+
   trait ExtraImplicits {
     /** Not in the standard scope due to the potential for divergence:
       * For instance `implicitly[Ordering[Any]]` diverges in its presence.
       */
     implicit def seqDerivedOrdering[CC[X] <: scala.collection.Seq[X], T](implicit ord: Ordering[T]): Ordering[CC[T]] =
```

Comment:
I'm wondering: why is the bound here not `<: Iterable[X]`?

Also, a comment why we have both this defintion and `implicit def Iterable` below would be very helpful.

Reply:
I don't know the reasons behind either of those things.

One of the problems with `implicit def Iterable` is that it only gives you an implicit `Ordering` if you need one for exactly `Iterable`, not for any subclass of `Iterable` - I suspect that's why the other one exists (although I don't know why we need both). The most I can say is that `seqDerivedOrdering` was added originally in https://github.com/scala/scala/commit/a887c83972cd1d4110d7ca17d982ee77f68ab0f0.

Reply:
@lrytz I've been thinking about this some more, and I'm not actually sure that we should have an implicit `Ordering` for `Iterable` - `Iterable`s are not guaranteed to have consistent element ordering, and since the `Ordering` implementation depends entirely on the element ordering, I think it would make sense to deprecate and remove it

Reply:
Yes, we should deprecate it and leave only `Seq`-based ordering.

Reply:
Do you mind if I make that a separate PR?

Reply:
No, that makes sense!